nntrainer/layers/grucell.cpp

```diff
@@ -0,0 +1,414 @@
+
+  Tensor &zrg_gates_derivatives =
+    context.getTensorGrad(wt_idx[GRUCellParams::zrg]);
+  zrg_gates_derivatives.reshape({max_timestep, 1, batch_size, NUM_GATE * unit});
+  Tensor zrg_gate_derivative = zrg_gates_derivatives.getBatchSlice(timestep, 1);
+  Tensor &weight_xh = context.getWeight(wt_idx[GRUCellParams::weight_xh]);
+  Tensor &outgoing_derivative = context.getOutgoingDerivative(SINGLE_INOUT_IDX);
+
+  zrg_gate_derivative.dot(weight_xh, outgoing_derivative, false, true);
+}
+
+void GRUCellLayer::calcGradient(RunLayerContext &context) {
+  const unsigned int unit = std::get<props::Unit>(grucell_props).get();
+  const float dropout_rate = std::get<props::DropOutRate>(grucell_props);
+  const unsigned int max_timestep = std::get<props::MaxTimestep>(grucell_props);
+  const unsigned int timestep = std::get<props::Timestep>(grucell_props);
+
+  Tensor &input = context.getInput(SINGLE_INOUT_IDX);
+  const TensorDim &input_dim = input.getDim();
+  const unsigned int batch_size = input_dim.batch();
+
+  Tensor &djdweight_xh =
+    context.getWeightGrad(wt_idx[GRUCellParams::weight_xh]);
+  Tensor &djdweight_hh =
+    context.getWeightGrad(wt_idx[GRUCellParams::weight_hh]);
+  Tensor &djdbias_ih = context.getWeightGrad(wt_idx[GRUCellParams::bias_h]);
+  Tensor &weight_hh = context.getWeight(wt_idx[GRUCellParams::weight_hh]);
+  Tensor &bias_h = context.getWeight(wt_idx[GRUCellParams::bias_h]);
+  Tensor bias_h_g = bias_h.getSharedDataTensor({unit}, 2 * unit);
+
+  Tensor djdw_zr_h =
+    djdweight_hh.getSharedDataTensor({unit, 2 * unit}, 0, false);
+  Tensor djdw_g_h =
+    djdweight_hh.getSharedDataTensor({unit, unit}, 2 * unit, false);
+  Tensor &hidden_states =
+    context.getTensor(wt_idx[GRUCellParams::hidden_state]);
+  hidden_states.reshape({max_timestep, 1, batch_size, unit});
+  Tensor hidden_state = hidden_states.getBatchSlice(timestep, 1);
+  Tensor &hidden_states_derivatives =
+    context.getTensorGrad(wt_idx[GRUCellParams::hidden_state]);
+  Tensor &incoming_derivative = context.getIncomingDerivative(SINGLE_INOUT_IDX);
+  Tensor &zrg_gates = context.getTensor(wt_idx[GRUCellParams::zrg]);
+  zrg_gates.reshape({max_timestep, 1, batch_size, NUM_GATE * unit});
+  Tensor zrg_gate = zrg_gates.getBatchSlice(timestep, 1);
+  Tensor &zrg_gates_derivatives =
+    context.getTensorGrad(wt_idx[GRUCellParams::zrg]);
+  zrg_gates_derivatives.reshape({max_timestep, 1, batch_size, NUM_GATE * unit});
+  Tensor zrg_gate_derivative = zrg_gates_derivatives.getBatchSlice(timestep, 1);
+
+  hidden_states_derivatives.reshape({max_timestep, 1, batch_size, unit});
+  Tensor hidden_state_derivative =
+    hidden_states_derivatives.getBatchSlice(timestep, 1);
+  hidden_state_derivative.reshape(incoming_derivative.getDim());
+  if (timestep + 1 == max_timestep) {
+    djdweight_xh.setZero();
+    djdweight_hh.setZero();
+    djdbias_ih.setZero();
+    hidden_state_derivative.copyData(incoming_derivative);
+    zrg_gates_derivatives.setZero();
```

Comment:
I think this is redundant. Please check.

Reply:
Good catch 👍